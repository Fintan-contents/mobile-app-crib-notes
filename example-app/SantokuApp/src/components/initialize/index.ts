export {WithInitializeContext, useInitializeContext} from './InitializeContext';
