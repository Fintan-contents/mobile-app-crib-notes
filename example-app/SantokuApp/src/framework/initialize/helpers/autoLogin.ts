import {AuthenticationService, isUnauthorizedError} from '../../authentication';
import {InitialDataError} from './initialDataError';

export const autoLogin = async () => {
  try {
    await AuthenticationService.autoLogin();
  } catch (e) {
    if (isUnauthorizedError(e)) {
      await AuthenticationService.clientLogout();
      throw e;
    }
    throw new InitialDataError(e);
  }
};
