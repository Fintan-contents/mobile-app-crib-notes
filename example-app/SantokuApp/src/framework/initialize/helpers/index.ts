export * from './firebase';
export * from './initialData';
export * from './initialDataError';
export * from './bundledMessage';
export * from './splashScreen';
export * from './updateRequired';
