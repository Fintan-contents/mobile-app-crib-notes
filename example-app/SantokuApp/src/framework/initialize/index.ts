export * from './withInitialData';
export * from './useAppInitializer';
