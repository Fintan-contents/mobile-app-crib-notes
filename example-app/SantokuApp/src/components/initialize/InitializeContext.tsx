import {createUseContextAndProvider} from 'framework/utilities';
import React, {useEffect, useMemo, useState} from 'react';

import {initialize, hideSplashScreen, NavigatorOptions} from './initialize';

type InitializeContextValue = {
  initialized: boolean;
  navigatorOptions: NavigatorOptions;
};

const [useInitializeContext, InitializeContextProvider] = createUseContextAndProvider<InitializeContextValue>();

const WithInitializeContext: React.FC = ({children}) => {
  const [error, setError] = useState<unknown>();
  const [initialized, setInitialized] = useState<boolean>(false);
  const [navigatorOptions, setNavigatorOptions] = useState<NavigatorOptions>({});
  const contextValue = useMemo(() => {
    return {
      initialized,
      navigatorOptions,
    };
  }, [initialized, navigatorOptions]);

  useEffect(() => {
    initialize()
      .then((navigatorOptions: NavigatorOptions) => {
        setNavigatorOptions(navigatorOptions);
        setInitialized(true);
        hideSplashScreen().catch(() => {});
      })
      .catch(e => {
        setError(e);
      });
  }, []);

  useEffect(() => {
    if (error) {
      // 初期化処理に失敗した場合はアプリをクラッシュ扱いで終了
      throw error;
    }
  }, [error]);

  if (initialized) {
    return <InitializeContextProvider value={contextValue}>{children}</InitializeContextProvider>;
  } else {
    return null;
  }
};

export {useInitializeContext, WithInitializeContext};
