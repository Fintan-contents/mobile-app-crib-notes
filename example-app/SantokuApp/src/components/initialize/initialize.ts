// import時の副作用を防ぐためにimport typeにしてます。
import type crashlyticsModule from '@react-native-firebase/crashlytics';
import {activateKeepAwake} from 'expo-keep-awake';
import * as SplashScreen from 'expo-splash-screen';
import {AppNavigatorOptions} from 'navigation/types';

import {BundledMessagesLoader, firebaseConfig, launchedId, loadMessages} from '../../framework';

type TermsAgreement = {
  hasAgreedValidTermsOfService: boolean;
  agreedTermsOfServiceVersion: string;
};

type InitialData = {
  terms?: TermsAgreement;
};

export const initialize = async () => {
  // 開発中は画面がスリープしないように設定
  if (__DEV__) {
    activateKeepAwake();
  }

  // Firebase Crashlyticsの初期化
  await initializeFirebaseCrashlyticsAsync();

  // アプリ内で使用するメッセージのロード
  await loadBundledMessagesAsync();

  // TODO: 通知をタップして起動した場合のパラメータ取得
  // TODO: ディープリンクから起動した場合のパラメータ取得
  // TODO: 強制アップデート対象バージョンかどうかの確認
  // TODO: キャッシュの削除

  // バックエンドから初期データを取得
  const initialData = await loadInitialDataAsync();

  // TODO: 読み込んだ初期データをFirebase Crashlyticsの設定に反映

  // 初期画面の決定
  return getInitialNavigatorOptions(initialData);
};

export const hideSplashScreen = async () => {
  // スプラッシュスクリーンを閉じる前に100ms程度は待たないと、コンポーネントのレンダリングが完了せず画面が一瞬白くなってしまいます。
  // expo-app-loadingでも200ms待っているので、ここでも200ms待つようにしています。
  // cf: https://github.com/expo/expo/blob/sdk-41/packages/expo-app-loading/src/AppLoadingNativeWrapper.tsx#L27
  await new Promise(resolve => setTimeout(resolve, 200));
  try {
    await SplashScreen.hideAsync();
  } catch (e) {
    // 既にSplashScreenが消えていた場合はそのまま継続
  }
};

const initializeFirebaseCrashlyticsAsync = async () => {
  if (!firebaseConfig.isDummy) {
    // requireした場合の型はanyとなってしまいESLintエラーが発生しますが無視します。
    // 型は、typeofで明示するようにします。
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const crashlytics = require('@react-native-firebase/crashlytics').default as typeof crashlyticsModule;
    // アプリの起動IDを設定
    try {
      const id = await launchedId();
      await crashlytics().setAttribute('launchedId', id);
    } catch (e) {
      throw new Error('Failed to get launchedId.');
    }
  }
};

const loadBundledMessagesAsync = async () => {
  try {
    await loadMessages(new BundledMessagesLoader());
  } catch (e) {
    // アプリにバンドルしているメッセージのロードは失敗しない想定
    throw new Error('Failed to load message.');
  }
};

const loadInitialDataAsync = async (): Promise<InitialData> => {
  // 利用規約同意状態を取得
  const terms = await getAccountMeTerms();
  return {
    terms,
  };
};

const getInitialNavigatorOptions = (initialData: InitialData): AppNavigatorOptions => {
  // TODO: 未認証状態ならログイン画面へ
  // TODO: チーム未参加状態ならチーム新規登録画面へ
  // TODO: 通知パラメータがあればそれに応じた画面へ
  // TODO: ディープリンクパラメータがあればそれに応じた画面へ
  if (!initialData.terms?.hasAgreedValidTermsOfService) {
    // 利用規約同意画面
    return {
      RootStackNav: {
        initialRouteName: 'TermsOfServiceAgreement',
      },
    };
  } else {
    // ホーム画面
    return {
      RootStackNav: {
        initialRouteName: 'AuthenticatedStackNav',
      },
      AuthenticatedStackNav: {
        initialRouteName: 'MainTabNav',
      },
      MainTabNav: {
        initialRouteName: 'HomeStackNav',
      },
      HomeStackNav: {
        initialRouteName: 'Home',
      },
    };
  }
};

// OpenAPI generatorで生成されたコードを導入するまでの一時的なMock
export const getAccountMeTerms = async () => {
  return await new Promise<TermsAgreement>(resolve =>
    setTimeout(
      () =>
        resolve({
          hasAgreedValidTermsOfService: false,
          agreedTermsOfServiceVersion: '1.0.0',
        }),
      100,
    ),
  );
};
