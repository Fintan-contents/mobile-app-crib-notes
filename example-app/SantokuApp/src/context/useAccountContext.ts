import {createUseContextAndProvider} from 'framework/utilities';
import {Account, TermsOfService, TermsOfServiceAgreementStatus} from 'generated/backend/model';

export type Terms = {
  termsAgreementStatus?: TermsOfServiceAgreementStatus;
  termsOfService?: TermsOfService;
};

export type AccountContext = {
  account?: Account;
  terms?: Terms;
  isLoggedIn: boolean;
};
export const [useAccountContext, AccountContextProvider] = createUseContextAndProvider<AccountContext>();
