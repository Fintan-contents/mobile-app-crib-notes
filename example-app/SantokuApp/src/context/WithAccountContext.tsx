import {AccountDataDependingComponent} from 'framework/initialize/withAccountData';
import React from 'react';

import {AccountContextProvider} from './useAccountContext';
import {
  AccountContextOperationProvider,
  accountContextReducer,
  useAccountOperation,
} from './useAccountContextOperation';

const WithAccountContext: AccountDataDependingComponent = ({accountData, children}) => {
  const account = accountData.account;
  const terms = accountData.terms;
  const initialAccountContext = account ? {account, terms, isLoggedIn: true} : {isLoggedIn: false};

  const [state, dispatch] = React.useReducer(accountContextReducer, initialAccountContext);
  const operations = useAccountOperation(initialAccountContext, dispatch);

  return (
    <AccountContextProvider value={state}>
      <AccountContextOperationProvider value={operations}>{children}</AccountContextOperationProvider>
    </AccountContextProvider>
  );
};

export {WithAccountContext};
