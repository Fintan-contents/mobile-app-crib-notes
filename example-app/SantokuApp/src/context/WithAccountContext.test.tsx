import {renderHook, WrapperComponent} from '@testing-library/react-hooks';
import {render} from '@testing-library/react-native';
import React from 'react';
import {Text} from 'react-native';

import {AccountData} from '../framework/initialize/helpers';
import {WithAccountContext} from './WithAccountContext';
import {useAccountContext} from './useAccountContext';

const wrapper: WrapperComponent<React.ProviderProps<AccountData>> = ({children, value}) => {
  return <WithAccountContext accountData={value}>{children}</WithAccountContext>;
};

describe('WithAccountContext', () => {
  const ChildComponent: React.FC = () => {
    return <Text testID="test">test</Text>;
  };

  it('WithAccountContextを子要素を含めて正常にrenderできること', () => {
    const accountData = {account: {accountId: '123456789', deviceTokens: []}};
    const withAccountContext = render(
      <WithAccountContext accountData={accountData}>
        <ChildComponent />
      </WithAccountContext>,
    );

    expect(withAccountContext.queryByTestId('test')).not.toBeNull();
    expect(withAccountContext).toMatchSnapshot();
  });

  it('初期データにAccountが存在している場合は、AccountContextのisLoggedInがtrueで取得できること', () => {
    const {result} = renderHook(() => useAccountContext(), {
      wrapper,
      initialProps: {value: {account: {accountId: '123456789', deviceTokens: []}}},
    });
    const accountContext = result.current;
    expect(accountContext).toEqual({account: {accountId: '123456789', deviceTokens: []}, isLoggedIn: true});
  });

  it('初期データにAccountが存在しない場合は、AccountContextのisLoggedInがfalseで取得できること', () => {
    const {result} = renderHook(() => useAccountContext(), {
      wrapper,
      initialProps: {value: {}},
    });
    const accountContext = result.current;
    expect(accountContext).toEqual({isLoggedIn: false});
  });
});
