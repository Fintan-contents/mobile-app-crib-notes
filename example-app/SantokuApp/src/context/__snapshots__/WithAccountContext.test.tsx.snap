// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithAccountContext WithAccountContextを子要素を含めて正常にrenderできること 1`] = `
<Text
  testID="test"
>
  test
</Text>
`;
